### PR TITLE
[https://nvbugs/5926823][fix] Propagate logprobs from prefill to decode in disagg

### PR DESCRIPTION
**Pull request by brb-nv**
## Description

This is a proposed fix for these bugs:
https://nvbugspro.nvidia.com/bug/5926823
https://nvbugspro.nvidia.com/bug/5926799

While both manifest differently, the core problem is that we don't transfer logprobs from prefill->decode in disagg.

Question for reviewers:
- Should we handle the case gracefully where logprobs are requested only on decode server but not on prefill? Or should we just throw saying logprobs should be requested on both prefill and decode?
- Any concerns about latency or size of the logprob payload? Since, logprobs seems to be top_k (unlike logits), I believe this should be ok?

## Test Coverage
```
$ pytest tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py::test_disaggregated_logprobs -s -v
$ pytest tests/unittest/disaggregated/test_openai_disagg_service.py::TestFirstGenLogProbsSerializeRoundtrip -s -v
```

- Repro script passes.
```
# 1. Kill any existing servers and make sure netstat returns empty.
pkill -f "cum_log_probs_repro.py" 2>/dev/null; pkill -f "trtllm-serve" 2>/dev/null; pkill -f "uvicorn" 2>/dev/null;
sleep 2
netstat -tlnp 2>/dev/null | grep 8000

# 2. Start prefill server on GPU 0 (logs to file).
cd /home/bbuddharaju/scratch/TensorRT-LLM
CUDA_VISIBLE_DEVICES=0 python3 cum_log_probs_repro.py prefill > output_prefill.log 2>&1 &

# 3. Wait for prefill server to be healthy.
echo "Waiting for prefill server..."
until curl -sf http://localhost:8000/health > /dev/null 2>&1; do sleep 5; done
echo "Prefill server ready!"

# 4. Run tests on GPU 1.
CUDA_VISIBLE_DEVICES=1 python3 cum_log_probs_repro.py test
```

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

`/bot [-h] ['run', 'kill', 'skip', 'reuse-pipeline'] ...`

Provide a user friendly way for developers to interact with a Jenkins server.

Run `/bot [-h|--help]` to print this help message.

See details below for each supported subcommand.

<details>

`run  [--reuse-test (optional)pipeline-id --disable-fail-fast --skip-test --stage-list "A10-PyTorch-1, xxx" --gpu-type "A30, H100_PCIe" --test-backend "pytorch, cpp" --add-multi-gpu-test --only-multi-gpu-test --disable-multi-gpu-test --post-merge --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx" --detailed-log --debug(experimental)]`

Launch build/test pipelines. All previously running jobs will be killed.

`--reuse-test (optional)pipeline-id ` *(OPTIONAL)* : Allow the new pipeline to reuse build artifacts and skip successful test stages from a specified pipeline or the last pipeline if no pipeline-id is indicated. If the Git commit ID has changed, this option will be always ignored. The DEFAULT behavior of the bot is to reuse build artifacts and successful test results from the last pipeline.

`--disable-reuse-test ` *(OPTIONAL)* : Explicitly prevent the pipeline from reusing build artifacts and skipping successful test stages from a previous pipeline. Ensure that all builds and tests are run regardless of previous successes.

`--disable-fail-fast ` *(OPTIONAL)* : Disable fail fast on build/tests/infra failures.

`--skip-test ` *(OPTIONAL)* : Skip all test stages, but still run build stages, package stages and sanity check stages. Note: Does **NOT** update GitHub check status.

`--stage-list "A10-PyTorch-1, xxx"` *(OPTIONAL)* : Only run the specified test stages. Examples: "A10-PyTorch-1, xxx". Note: Does **NOT** update GitHub check status.

`--gpu-type "A30, H100_PCIe"` *(OPTIONAL)* : Only run the test stages on the specified GPU types. Examples: "A30, H100_PCIe". Note: Does **NOT** update GitHub check status.

`--test-backend "pytorch, cpp"` *(OPTIONAL)* : Skip test stages which don't match the specified backends. Only support [pytorch, cpp, tensorrt, triton]. Examples: "pytorch, cpp" (does not run test stages with tensorrt or triton backend). Note: Does **NOT** update GitHub pipeline status.

`--only-multi-gpu-test ` *(OPTIONAL)* : Only run the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--disable-multi-gpu-test ` *(OPTIONAL)* : Disable the multi-GPU tests. Note: Does **NOT** update GitHub check status.

`--add-multi-gpu-test ` *(OPTIONAL)* : Force run the multi-GPU tests in addition to running L0 pre-merge pipeline.

`--post-merge ` *(OPTIONAL)* : Run the L0 post-merge pipeline instead of the ordinary L0 pre-merge pipeline.

`--extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx"` *(OPTIONAL)* : Run the ordinary L0 pre-merge pipeline and specified test stages. Examples: --extra-stage "H100_PCIe-TensorRT-Post-Merge-1, xxx".

`--detailed-log ` *(OPTIONAL)* : Enable flushing out all logs to the Jenkins console. This will significantly increase the log volume and may slow down the job.

`--debug ` *(OPTIONAL)* : **Experimental feature**. Enable access to the CI container for debugging purpose. Note: Specify exactly one stage in the `stage-list` parameter to access the appropriate container environment. Note: Does **NOT** update GitHub check status.

For guidance on mapping tests to stage names, see `docs/source/developer-guide/ci-overview.md`
and the `scripts/test_to_stage_mapping.py` helper.

### kill

`kill  `

Kill all running builds associated with pull request.

### skip

`skip --comment COMMENT `

Skip testing for latest commit on pull request. `--comment "Reason for skipping build/test"` is required. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

### reuse-pipeline

`reuse-pipeline `

Reuse a previous pipeline to validate current commit. This action will also kill all currently running builds associated with the pull request. IMPORTANT NOTE: This is dangerous since lack of user care and validation can cause top of tree to break.

</details>

## Release Notes

* **New Features**
  * Added support for cumulative log probabilities in disaggregated (separate prefill/decode) serving mode.
  * Log probabilities can now be transferred from prefill to decode stages.

* **Bug Fixes**
  * Improved robustness of log probability attribute access with explicit guard checks to prevent errors.

* **Tests**
  * Added comprehensive tests for log probability workflow in disaggregated serving scenarios.

### CODE REVIEW

File: tensorrt_llm/disaggregated_params.py [hunk 1]
```diff
@@ -37,6 +39,7 @@ class DisaggregatedParams:
     request_type: Optional[str] = None
     # P-D Disaggregated Params
     first_gen_tokens: Optional[List[int]] = None
+    first_gen_log_probs: Optional[List] = None
```
[the remark this answers is not in the document]
Comment (yibinl-nvidia) [reply tensorrt_llm/disaggregated_params.py]: Hi @brb-nv , pickle.load is inherently unsafe so we discourage its usage. If you really want to use pickle.load functionality here and know the dtype that are going to be deserialized, I recommend checking https://github.com/NVIDIA/TensorRT-LLM/pull/10622/changes to use restricted pickler.

File: tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py [hunk 2]
```diff
@@ -510,5 +510,104 @@ def test_disaggregated_spec_dec_batch_slot_limit(model, spec_dec_model_path,
             print("All workers terminated.")
 
 
+@pytest.mark.parametrize("model", ["TinyLlama-1.1B-Chat-v1.0"])
+@pytest.mark.parametrize("generation_overlap", [False, True])
+def test_disaggregated_logprobs(model, generation_overlap):
```
Comment (StanleySun639) [inline tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py]: Could you add the new test to L0 or QA test [list](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/qa/llm_function_core.txt) to get tested.

**Review by StanleySun639 — CHANGES_REQUESTED** [review]

File: tensorrt_llm/disaggregated_params.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tabrizian) [reply tensorrt_llm/disaggregated_params.py]: The main reason I was suggesting this was to avoid having the user update their API to get log probs (e.g. Dynamo has to do this additional disagg-only step for logprobs and perhaps logits too in the future).

[the remark this answers is not in the document]
Comment (Tabrizian) [reply tensorrt_llm/disaggregated_params.py]: Based on offline discussion, having a new field is fine and doesn't change the API. Resolving.

[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/disaggregated_params.py]: Thank you for the kind inputs everyone! Updated change: https://github.com/NVIDIA/TensorRT-LLM/pull/11727/commits/3ee8d7d164f8b7133aed12c6a486eb60c14b50b6

File: tensorrt_llm/executor/result.py [hunk 3]
```diff
@@ -289,16 +290,33 @@ def _handle_sequence(self,
                 output.logprobs += response_tensors.log_probs[src_idx]
 
             # overcome some WAR in the cpp executor
-            if finish_reasons[
-                    src_idx] != tllm.FinishReason.CANCELLED and self.use_trtllm_sampler:
```
Comment (pcastonguay) [inline tensorrt_llm/executor/result.py]: Do you know we have that `and self.use_trtllm_sampler`?

**Review by Tabrizian — APPROVED** [review]
one minor comment, otherwise looks good to me.

File: tensorrt_llm/executor/result.py [hunk 4]
```diff
@@ -289,16 +290,33 @@ def _handle_sequence(self,
...
+            if finish_reasons[src_idx] != tllm.FinishReason.CANCELLED:
                 if len(output.logprobs) > output.length:
                     # LlmResult holds a reference to LogProbStorage, which may be updated by the worker before the result is serialized.
                     # Therefore, we treat extra logprobs/logits as expected and only consume what's needed.
                     output.logprobs = output.logprobs[:output.length]
-            assert len(
-                output.logprobs
-            ) == output.length, f"logprobs length: {len(output.logprobs)} != output.length: {output.length}"
+
+            if finish_reasons[src_idx] != tllm.FinishReason.CANCELLED:
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/executor/result.py]: Combined here. https://github.com/NVIDIA/TensorRT-LLM/pull/11727/commits/dfbb0bf08f3056942da2df65185e71fad116c9b6

File: tensorrt_llm/executor/result.py [hunk 5]
```diff
@@ -289,16 +290,33 @@ def _handle_sequence(self,
...
+                        f"output.length - 1: {output.length - 1}")
+                    if len(output.logprobs) < output.length:
+                        logger.warning(
+                            "Disaggregated serving: logprobs for the first "
+                            "generated token were not transferred from the "
+                            "context server. The response will contain %d "
+                            "logprob entries instead of %d. Ensure "
+                            "first_gen_log_probs is propagated in "
+                            "DisaggregatedParams to avoid incomplete "
+                            "results.", len(output.logprobs), output.length)
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/executor/result.py]: Updated the comment to propose an action to the user. https://github.com/NVIDIA/TensorRT-LLM/pull/11727/commits/dfbb0bf08f3056942da2df65185e71fad116c9b6

File: tensorrt_llm/executor/result.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (brb-nv) [reply tensorrt_llm/executor/result.py]: Done. https://github.com/NVIDIA/TensorRT-LLM/pull/11727/commits/dfbb0bf08f3056942da2df65185e71fad116c9b6

**Review by pcastonguay — APPROVED** [review]

**Review by StanleySun639 — APPROVED** [review]

**Review by chuangz0 — APPROVED** [review]

File: tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py [hunk 6]
```diff
@@ -510,5 +510,111 @@ def test_disaggregated_spec_dec_batch_slot_limit(model, spec_dec_model_path,
             print("All workers terminated.")
 
 
+@pytest.mark.parametrize("model", ["TinyLlama-1.1B-Chat-v1.0"])
```
[the remark this answers is not in the document]
Comment (brb-nv) [reply tests/integration/defs/disaggregated/test_disaggregated_single_gpu.py]: As discussed offline, I'll address this in a different MR: https://github.com/NVIDIA/TensorRT-LLM/pull/11767
